In [1]:
%load_ext autoreload
%load_ext autotime

time: 433 µs (started: 2023-07-03 19:27:48 -07:00)


In [2]:
import sys
sys.path.append("../")
import warnings
warnings.filterwarnings('ignore')
from bokeh.io import output_notebook
output_notebook()

Loading BokehJS ...

time: 1.38 s (started: 2023-07-03 19:27:49 -07:00)


In [3]:
import dask.dataframe as dd
import numpy as np
import pandas as pd


time: 684 ms (started: 2023-07-03 19:27:50 -07:00)


In [4]:
%autoreload 2

from wisio.recorder import RecorderAnalyzer

log_dir = "/p/vast1/iopp/recorder_app_logs/cm1/nodes-32/workflow-4/_parquet"

recorder_analyzer = RecorderAnalyzer(
    cluster_manager_args=dict(
        force_local=True,
        cluster_settings=dict(
            cores=4,
            dashboard_port=3646,
            local_directory="/var/tmp/dask-recorder",
            log_file="%J.log",
            worker_queue="pdebug"
        )
    ),
    working_dir='.recorder',
    debug=True
) 

recorder_analyzer

distributed.diskutils - INFO - Found stale lock file and directory '/var/tmp/dask-recorder/cluster-0-local/dask-worker-space/worker-jbwwyfw1', purging
distributed.diskutils - INFO - Found stale lock file and directory '/var/tmp/dask-recorder/cluster-0-local/dask-worker-space/worker-ykrs7zij', purging
distributed.diskutils - INFO - Found stale lock file and directory '/var/tmp/dask-recorder/cluster-0-local/dask-worker-space/worker-fu0qkw20', purging
distributed.diskutils - INFO - Found stale lock file and directory '/var/tmp/dask-recorder/cluster-0-local/dask-worker-space/worker-yzva5147', purging


time: 2.26 s (started: 2023-07-03 19:27:51 -07:00)


In [5]:
recorder_analyzer.cluster_manager.clients[0]

<Client: 'tcp://192.168.133.252:37329' processes=4 threads=36>

time: 51.3 ms (started: 2023-07-03 19:27:53 -07:00)


In [14]:
%autoreload 2

res = recorder_analyzer.analyze_parquet(log_dir=log_dir, checkpoint=False)

first_ix 0.7869744300842285
first_ix 0.08024907112121582
first_ix 0.07010746002197266
first_ix 0.14856934547424316
first_ix 0.14993000030517578
first_ix 0.17853498458862305
first_ix 0.14509224891662598
first_ix 0.1342768669128418
first_ix 0.1290452480316162
first_ix 0.21705174446105957
first_ix 0.21612858772277832
first_ix 0.23006725311279297
first_ix 0.2279045581817627
first_ix 0.20815587043762207
first_ix 0.2105855941772461
first_ix 0.17177772521972656
first_ix 0.16691255569458008
first_ix 0.16660499572753906
first_ix 0.17094182968139648
first_ix 0.178572416305542
time: 11.3 s (started: 2023-07-03 19:37:00 -07:00)


In [15]:
%autoreload 2

res = recorder_analyzer.analyze_parquet(log_dir=log_dir, checkpoint=False)

first_ix 0.0023756027221679688
first_ix 0.0022530555725097656
first_ix 0.0024247169494628906
first_ix 0.002527952194213867
first_ix 0.00407719612121582
first_ix 0.0022552013397216797
first_ix 0.0026960372924804688
first_ix 0.002290964126586914
first_ix 0.002191305160522461
first_ix 0.0024712085723876953
first_ix 0.0024225711822509766
first_ix 0.0024394989013671875
first_ix 0.002359151840209961
first_ix 0.0024285316467285156
first_ix 0.002408266067504883
first_ix 0.002232074737548828
first_ix 0.002181529998779297
first_ix 0.0022630691528320312
first_ix 0.00235748291015625
first_ix 0.002247333526611328
time: 7.23 s (started: 2023-07-03 19:37:46 -07:00)


In [16]:
main_view, views, bottlenecks = res

time: 21.5 ms (started: 2023-07-03 19:37:57 -07:00)


In [17]:
views[('trange',)].head()

duration_sum  \
trange file_name                                          proc_name                                          
5      /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#1000#35184372411024      2.151723   
                                                          app1#localhost#1040#35184372411024      1.952563   
                                                          app1#localhost#1080#35184372411024      2.116145   
                                                          app1#localhost#1120#35184372411024      2.116290   
                                                          app1#localhost#1160#35184372411024      2.026798   

                                                                                              index_count  \
trange file_name                                          proc_name                                         
5      /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#1000#35184372411024            1   
                                                          app1#localhost#1040#35184372411024            1   
                                                          app1#localhost#1080#35184372411024            1   
                                                          app1#localhost#1120#35184372411024            1   
                                                          app1#localhost#1160#35184372411024            1   

                                                                                              size_min  \
trange file_name                                          proc_name                                      
5      /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#1000#35184372411024         0   
                                                          app1#localhost#1040#35184372411024         0   
                                                          app1#localhost#1080#35184372411024         0   
                                                          app1#localhost#1120#35184372411024         0   
                                                          app1#localhost#1160#35184372411024         0   

                                                                                              size_max  \
trange file_name                                          proc_name                                      
5      /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#1000#35184372411024         0   
                                                          app1#localhost#1040#35184372411024         0   
                                                          app1#localhost#1080#35184372411024         0   
                                                          app1#localhost#1120#35184372411024         0   
                                                          app1#localhost#1160#35184372411024         0   

                                                                                              size_sum  \
trange file_name                                          proc_name                                      
5      /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#1000#35184372411024         0   
                                                          app1#localhost#1040#35184372411024         0   
                                                          app1#localhost#1080#35184372411024         0   
                                                          app1#localhost#1120#35184372411024         0   
                                                          app1#localhost#1160#35184372411024         0   

                                                                                              read_time  \
trange file_name                                          proc_name                                       
5      /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#1000#35184372411024        0.0   
         

time: 201 ms (started: 2023-07-03 19:37:58 -07:00)


In [10]:
from wisio.rules import Rule
from wisio._recorder.rules import RecorderRuleEngine

    # app_name=('app_name', 'trange', 'file_name'),
    # file_dir=('file_dir', 'proc_name', 'trange'),
    # file_name=('file_name', 'proc_name', 'trange'),
    # file_regex=('file_regex', 'proc_name', 'trange'),
    # node_name=('node_name', 'trange', 'file_name'),
    # proc_name=('proc_name', 'trange', 'file_name'),
    # rank=('rank', 'trange', 'file_name'),
    # trange=('trange', 'proc_name', 'file_name'),

rules = dict(
    app_name=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    file_dir=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    file_name=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    file_regex=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    node_name=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    proc_name=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    rank=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
    trange=[Rule.BOTT_METADATA_ACCESS, Rule.BOTT_SMALL_READS, Rule.BOTT_SMALL_WRITES],
)

rule_engine = RecorderRuleEngine(rules=rules)

rule_engine


time: 13.4 ms (started: 2023-07-03 19:30:28 -07:00)


In [11]:
%autoreload 2

characteristics = rule_engine.process_characteristics(view=main_view)
characteristics


{<Rule.CHAR_IO_TIME: 118>: RuleResult(data_dict=None, description='I/O Time', detail_list=['Read - 0.00 seconds (0.05%)', 'Write - 0.58 seconds (14.08%)', 'Metadata - 3.53 seconds (85.89%)'], reasons=None, rule=<Rule.CHAR_IO_TIME: 118>, value=4.1151000122494, value_fmt='4.12 seconds'),
 <Rule.CHAR_IO_COUNT: 111>: RuleResult(data_dict=None, description='I/O Ops', detail_list=['Read - 1,282 ops (4.67%)', 'Write - 2,303 ops (8.39%)', 'Metadata - 23,878 ops (86.95%)'], reasons=None, rule=<Rule.CHAR_IO_COUNT: 111>, value=27463, value_fmt='27,463 ops'),
 <Rule.CHAR_IO_SIZE: 116>: RuleResult(data_dict=None, description='I/O Size', detail_list=['Read - 20.03 GB (94.51%)', 'Write - 1.15 GB (5.41%)'], reasons=None, rule=<Rule.CHAR_IO_SIZE: 116>, value=22756918836, value_fmt='21.19 GB'),
 <Rule.CHAR_READ_XFER_SIZE: 132>: RuleResult(data_dict=None, description='Read Xfer', detail_list=['4-16 MB - 1,282 ops (100.00%)'], reasons=None, rule=<Rule.CHAR_READ_XFER_SIZE: 132>, value=(16777216, 16777216),

time: 197 ms (started: 2023-07-03 19:30:29 -07:00)


In [12]:
from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich.tree import Tree

# Create a new Table object from Rich library
table = Table(box=None, show_header=False)

# Add columns to the table for the key and value
table.add_column(style="cyan")
table.add_column()

# Add each key-value pair to the table as a row
for rule, result in characteristics.items():
    if result.detail_list is None:
        table.add_row(result.description, result.value_fmt)
    else:
        detail_tree = Tree(result.value_fmt)
        for detail in result.detail_list:
            detail_tree.add(detail)
        table.add_row(result.description, detail_tree)

# Create a Console object from Rich library
console = Console()

# Print the table with Rich formatting
console.print(Panel(table, title='Summary'))


╭──────────────────────────────────────────────────── Summary ────────────────────────────────────────────────────╮
│  I/O Time        4.12 seconds                                                                                   │
│                  ├── Read - 0.00 seconds (0.05%)                                                                │
│                  ├── Write - 0.58 seconds (14.08%)                                                              │
│                  └── Metadata - 3.53 seconds (85.89%)                                                           │
│  I/O Ops         27,463 ops                                                                                     │
│                  ├── Read - 1,282 ops (4.67%)                                                                   │
│                  ├── Write - 2,303 ops (8.39%)                                                                  │
│                  └── Metadata - 23,878 ops (86.95%)                                                             │
│  I/O Size        21.19 GB                                                                                       │
│                  ├── Read - 20.03 GB (94.51%)                                                                   │
│                  └── Write - 1.15 GB (5.41%)                                                                    │
│  Read Xfer       16 MB - 1,282 ops                                                                              │
│                  └── 4-16 MB - 1,282 ops (100.00%)                                                              │
│  Write Xfer      4 KB-16 MB - 2,254 ops                                                                         │
│                  ├── <4 KB - 397 ops (17.61%)                                                                   │
│                  ├── 4-16 KB - 1,092 ops (48.45%)                                                               │
│                  ├── 16-64 KB - 722 ops (32.03%)                                                                │
│                  ├── 64-256 KB - 1 ops (0.04%)                                                                  │
│                  └── 4-16 MB - 42 ops (1.86%)                                                                   │
│  Apps            1 apps                                                                                         │
│                  └── app1 - 4.12 s (100.00%) - 20.03/1.15 GB R/W (94.51/5.41%) - 27,463 ops (100.00%)           │
│  Nodes           1 nodes                                                                                        │
│                  └── localhost - 4.12 s (100.00%) - 20.03/1.15 GB R/W (94.51/5.41%) - 27,463 ops (100.00%)      │
│  Files           775 files                                                                                      │
│                  ├── Shared: 775 files (100.00%)                                                                │
│                  └── FPP: 0 files (0.00%)                                                                       │
│  Access Pattern  100.00% Sequential - 0.00% Random                                                              │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

time: 200 ms (started: 2023-07-03 19:30:29 -07:00)


In [13]:
%autoreload 2

processed_bottlenecks = rule_engine.process_bottlenecks(bottlenecks=bottlenecks, characteristics=characteristics, threshold=0.01)
processed_bottlenecks

{('trange',): {<Rule.BOTT_METADATA_ACCESS: 207>: {5: RuleResult(data_dict={'llc': {'duration_sum': 2.191842794418335, 'index_count': 32, 'size_min': 0, 'size_max': 0, 'size_sum': 0, 'read_time': 0.0, 'write_time': 0.0, 'metadata_time': 2.191842794418335, 'read_size': 0, 'write_size': 0, 'metadata_size': 0, 'read_count': 0, 'write_count': 0, 'metadata_count': 32, 'read_min': 0, 'read_max': 0, 'write_min': 0, 'write_max': 0, 'metadata_min': 0, 'metadata_max': 0, 'data_count': 0, 'data_size': 0, 'data_time': 0.0, 'sequential_time': 0.0, 'random_time': 0.0, 'sequential_size': 0, 'random_size': 0, 'sequential_count': 0, 'random_count': 0, 'close_time': 0.0, 'open_time': 2.191842794418335, 'seek_time': 0.0, 'stat_time': 0.0, 'close_count': 0, 'open_count': 32, 'seek_count': 0, 'stat_count': 0, 'duration_pero': 0.37435507603975887, 'duration_perr': 1.0, 'duration_score': 'medium', 'duration_th': 0.25}, 'proc_name_list': ['app1#localhost#0#35184372411024', 'app1#localhost#1000#35184372411024',

time: 4.07 s (started: 2023-07-03 19:30:30 -07:00)


In [ ]:
from rich.tree import Tree
from rich.console import Console

# create a tree object
tree = Tree("Bottlenecks")

for view_key, bottleneck_dict in processed_bottlenecks.items():
    branch_name = '_'.join(view_key) if isinstance(view_key, tuple) else view_key
    branch = tree.add(branch_name)
    for rule, rule_results in bottleneck_dict.items():
        subbranch = branch.add(rule.name)
        for bottleneck_key, result in rule_results.items():
            subbranch.add(result.description)


# create a console object and print the tree
console = Console()
console.print(tree)


In [18]:
from rich.table import Table
from rich.console import Console

# define the data as a list of tuples
data = [
    ("Root", None),
    ("Branch 1", "Root"),
    ("Leaf 1.1", "Branch 1"),
    ("Leaf 1.2", "Branch 1"),
    ("Branch 2", "Root"),
    ("Leaf 2.1", "Branch 2"),
    ("Leaf 2.2", "Branch 2"),
]

# create a table object and add columns
table = Table(show_header=False, box=None)
table.add_column(justify="left")
table.add_column(justify="left")

# add rows to the table
for row in data:
    if row[1] is None:
        table.add_row(row[0])
    else:
        table.add_row("", row[0])

# print the table
console = Console()
console.print(table)


Root           
       Branch 1 
       Leaf 1.1 
       Leaf 1.2 
       Branch 2 
       Leaf 2.1 
       Leaf 2.2

time: 7.39 ms (started: 2023-03-22 22:09:00 -07:00)


'/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat': 
RuleResult(rule=<Rule.METADATA_ACCESS_ISSUE: 21>, 
           
           description="'/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat' is 
           accessed by 32 process(es) during the 4th second(s) and spent 100.00% (2.19 seconds) of its I/O time on the 
           'open' operation(s).", 
        
    reasons=[RuleReason(description='Metadata time is 100.00% (2.19 seconds) of I/O time', value=1.0)])

- observation: '/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat' is 
           accessed by 32 process(es) during the 4th second(s)
- reason for bottleneck selection: has an I/O time of X which in Y% of overall I/O time of the workload.
- reason1: Overall % of I/O time is spent on metadata access, specifically % on 'open' ... and % on others. 

SMALL_IO_ACCESS

description = (
            f"'{ix}' is accessed by process(es) {join_with_and(processes_fmt)} "
            f"during the {join_with_and(values=trange_intervals)}th second(s) "
            f"with a transfer size smaller than 256KB ({max_xfer_size / 1024:.2f}KB)."
        )

- observation: '{ix}' is accessed by process(es) {join_with_and(processes_fmt)} during the {join_with_and(values=trange_intervals)}th second(s)
- reason for bottleneck: has an I/O time of X which in Y% of overall I/O time of the workload.
- reason1: Average I/O accesses on this file are smaller than 256KB, which is a small data access for GPFS. 
           Specifically ... (histogram or frequency/read or write)
- reason2: If metadata access > 50% ... 
TODO : check whether the file is on GPFS or not

In [15]:
files = ['/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat','/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat']
files

['/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat',
 '/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/cm1out_000001_s.dat']

time: 3.66 ms (started: 2023-03-16 11:49:16 -07:00)


In [16]:
all(['/gpfs' in f for f in files])

True

time: 3.63 ms (started: 2023-03-16 11:49:39 -07:00)


In [16]:
trange_hl_view = bottlenecks[('trange',)]['high_level_view'].compute()
trange_hl_view


duration_sum  index_count  size_min  size_max     size_sum  read_time  \
trange                                                                          
1           0.001612            6         0  16777216     16777216   0.001470   
3           0.020756         7673         0  16777216  21441282048   0.001981   
4           0.004128           58         0  16777216     50456557   0.001317   
5           2.191843           32         0         0            0   0.000000   
6           0.011320          384         0  16777216     33560897   0.000000   
...              ...          ...       ...       ...          ...        ...   
664         0.001761           40      5530     97648       227631   0.000000   
666         0.001675           40      5530     97920       227903   0.000000   
667         0.027776          142         0     98192       104505   0.000000   
668         0.277307          909         0  16777216    146154226   0.000000   
669         0.017113          133         0  16777216     80746464   0.000000   

        write_time  metadata_time    read_size  write_size  ...  \
trange                                                      ...   
1         0.000000       0.000142     16777216           0  ...   
3         0.000000       0.020747  21441282048           0  ...   
4         0.000201       0.002610     50331648      124909  ...   
5         0.000000       2.191843            0           0  ...   
6         0.006047       0.005273            0    33560897  ...   
...            ...            ...          ...         ...  ...   
664       0.000411       0.001350            0      130255  ...   
666       0.000377       0.001298            0      130255  ...   
667       0.000191       0.027585            0        6465  ...   
668       0.027584       0.277307            0   146154226  ...   
669       0.016762       0.000350            0    80746464  ...   

        sequential_count  random_count  close_time  open_time  seek_time  \
trange                                                                     
1                      1             0    0.000009   0.000122   0.000000   
3                   1278             0    0.000012   0.020717   0.000003   
4                      5             0    0.000083   0.002489   0.000002   
5                      0             0    0.000000   2.191843   0.000000   
6                      6             0    0.000065   0.005128   0.000004   
...                  ...           ...         ...        ...        ...   
664                    6             0    0.000173   0.000645   0.000012   
666                    6             0    0.000169   0.000627   0.000011   
667                    4             0    0.000051   0.027195   0.000011   
668                   19             0    0.000479   0.277214   0.000014   
669                    8             0    0.000350   0.000000   0.000000   

        stat_time  duration_pero  duration_perr  duration_score  duration_th  
trange                                                                        
1        0.000011       0.000392       0.000275            none        0.000  
3        0.000020       0.005044       0.003545         trivial        0.001  
4        0.000027       0.001003       0.000705         trivial        0.001  
5        0.000000       0.532634       0.374355            high        0.500  
6        0.000057       0.002751       0.001933         trivial        0.001  
...           ...            ...            ...             ...          ...  
664      0.000054       0.000428       0.000301            none        0.000  
666      0.000059       0.000407       0.000286            none        0.000  
667      0.000051       0.006750       0.004744         trivial        0.001  
668      0.000135       0.067388       0.047363        very low        0.010  
669      0.000000       0.004159       0.002923         trivial        0.001  

[393 rows x 31 columns]

time: 305 ms (started: 2023-03-16 09:20:07 -07:00)


In [17]:
trange_hl_view['duration_th'].value_counts().sort_index()


0.000    292
0.001     89
0.010     11
0.500      1
Name: duration_th, dtype: int64

time: 4.34 ms (started: 2023-03-16 09:20:08 -07:00)


In [18]:
trange_hl_view.query('duration_th >= 0.01')


NameError: name 'trange_hl_view' is not defined

time: 15.9 ms (started: 2023-03-16 16:21:50 -07:00)


In [19]:
trange_ml_view = bottlenecks[('trange',)]['mid_level_view'].compute()
trange_ml_view


duration_sum  index_count  size_min  \
trange proc_name                                                                
1      app1#localhost#17#35184372411024       0.001612            6         0   
3      app1#localhost#0#35184372411024        0.020756           11         0   
       app1#localhost#1#35184372411024        0.000040            6         0   
       app1#localhost#10#35184372411024       0.000037            6         0   
       app1#localhost#100#35184372411024      0.000040            6         0   
...                                                ...          ...       ...   
669    app1#localhost#800#35184372411024      0.000044            4         0   
       app1#localhost#840#35184372411024      0.000044            4         0   
       app1#localhost#880#35184372411024      0.000043            4         0   
       app1#localhost#920#35184372411024      0.000042            4         0   
       app1#localhost#960#35184372411024      0.000047            4         0   

                                          size_max  size_sum  read_time  \
trange proc_name                                                          
1      app1#localhost#17#35184372411024   16777216  16777216   0.001470   
3      app1#localhost#0#35184372411024    16777216  16777216   0.000009   
       app1#localhost#1#35184372411024    16777216  16777216   0.000009   
       app1#localhost#10#35184372411024   16777216  16777216   0.000008   
       app1#localhost#100#35184372411024  16777216  16777216   0.000010   
...                                            ...       ...        ...   
669    app1#localhost#800#35184372411024         0         0   0.000000   
       app1#localhost#840#35184372411024         0         0   0.000000   
       app1#localhost#880#35184372411024         0         0   0.000000   
       app1#localhost#920#35184372411024         0         0   0.000000   
       app1#localhost#960#35184372411024         0         0   0.000000   

                                          write_time  metadata_time  \
trange proc_name                                                      
1      app1#localhost#17#35184372411024          0.0       0.000142   
3      app1#localhost#0#35184372411024           0.0       0.020747   
       app1#localhost#1#35184372411024           0.0       0.000031   
       app1#localhost#10#35184372411024          0.0       0.000029   
       app1#localhost#100#35184372411024         0.0       0.000030   
...                                              ...            ...   
669    app1#localhost#800#35184372411024         0.0       0.000044   
       app1#localhost#840#35184372411024         0.0       0.000044   
       app1#localhost#880#35184372411024         0.0       0.000043   
       app1#localhost#920#35184372411024         0.0       0.000042   
       app1#localhost#960#35184372411024         0.0       0.000047   

                                          read_size  write_size  ...  \
trange proc_name                                                 ...   
1      app1#localhost#17#35184372411024    16777216           0  ...   
3      app1#localhost#0#35184372411024     16777216           0  ...   
       app1#localhost#1#35184372411024     16777216           0  ...   
       app1#localhost#10#35184372411024    16777216           0  ...   
       app1#localhost#100#35184372411024   16777216           0  ...   
...                                             ...         ...  ...   
669    app1#localhost#800#35184372411024          0           0  ...   
       app1#localhost#840#35184372411024          0           0  ...   
       app1#localhost#880#35184372411024          0           0  ...   
       app1#localhost#920#35184372411024          0           0  ...   
       app1#localhost#960#35184372411024          0           0  ...   

                                          sequential_count  random_count  \
trange proc_name                                                           

time: 246 ms (started: 2023-03-16 09:20:08 -07:00)


In [20]:
trange_ml_view['duration_th'].value_counts().sort_index()


0.000    1839
0.001     128
0.010     198
0.250      13
0.500      19
Name: duration_th, dtype: int64

time: 4.08 ms (started: 2023-03-16 09:20:09 -07:00)


In [21]:
trange_ll_view = bottlenecks[('trange',)]['low_level_view'].compute()
trange_ll_view


duration_sum  \
trange proc_name                         file_name                                                          
1      app1#localhost#17#35184372411024  /dev/null                                               0.000003   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.001609   
3      app1#localhost#0#35184372411024   /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000034   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.020722   
       app1#localhost#1#35184372411024   /dev/null                                               0.000003   
...                                                                                                   ...   
669    app1#localhost#920#35184372411024 /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000005   
       app1#localhost#960#35184372411024 /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000033   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000005   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000004   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000005   

                                                                                             index_count  \
trange proc_name                         file_name                                                         
1      app1#localhost#17#35184372411024  /dev/null                                                     1   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            5   
3      app1#localhost#0#35184372411024   /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            5   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            6   
       app1#localhost#1#35184372411024   /dev/null                                                     1   
...                                                                                                  ...   
669    app1#localhost#920#35184372411024 /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            1   
       app1#localhost#960#35184372411024 /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            1   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            1   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            1   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            1   

                                                                                             size_min  \
trange proc_name                         file_name                                                      
1      app1#localhost#17#35184372411024  /dev/null                                                  0   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  16777216   
3      app1#localhost#0#35184372411024   /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  16777216   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   
       app1#localhost#1#35184372411024   /dev/null                                                  0   
...                                                                                               ...   
669    app1#localhost#920#35184372411024 /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   
       app1#localhost#960#35184372411024 /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   
                                         /p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   
          

time: 234 ms (started: 2023-03-16 09:20:09 -07:00)


In [17]:
trange_ll_view['duration_th'].value_counts().sort_index()


0.000    6412
0.001      78
0.010     204
0.250      13
0.500      19
Name: duration_th, dtype: int64

time: 5.52 ms (started: 2023-03-15 13:02:46 -07:00)


In [19]:
proc_name_hl_view = bottlenecks[('proc_name',)]['high_level_view'].compute()
proc_name_hl_view


duration_sum  index_count  size_min  \
proc_name                                                                 
app1#localhost#0#35184372411024         4.115100        15028         0   
app1#localhost#1000#35184372411024      3.061782          159         0   
app1#localhost#1038#35184372411024      0.000710            6         0   
app1#localhost#1040#35184372411024      3.071674          159         0   
app1#localhost#1068#35184372411024      0.000701            6         0   
app1#localhost#1080#35184372411024      2.932791          159         0   
app1#localhost#1111#35184372411024      0.002125            6         0   
app1#localhost#112#35184372411024       0.000855            6         0   
app1#localhost#1120#35184372411024      2.950001          164         0   
app1#localhost#1155#35184372411024      0.001996            6         0   
app1#localhost#1160#35184372411024      2.781130          159         0   
app1#localhost#1196#35184372411024      0.001549            6         0   
app1#localhost#120#35184372411024       2.984170          158         0   
app1#localhost#1200#35184372411024      2.968496          160         0   
app1#localhost#1224#35184372411024      0.001246            6         0   
app1#localhost#1240#35184372411024      3.239430          158         0   
app1#localhost#148#35184372411024       0.000896            6         0   
app1#localhost#155#35184372411024       0.000762            6         0   
app1#localhost#160#35184372411024       2.787134          158         0   
app1#localhost#17#35184372411024        0.001612            6         0   
app1#localhost#175#35184372411024       0.000654            6         0   
app1#localhost#200#35184372411024       3.238349          162         0   
app1#localhost#240#35184372411024       2.829355          160         0   
app1#localhost#255#35184372411024       0.001741            6         0   
app1#localhost#280#35184372411024       2.957242          158         0   
app1#localhost#296#35184372411024       0.000536            6         0   
app1#localhost#320#35184372411024       3.008698          158         0   
app1#localhost#342#35184372411024       0.000572            6         0   
app1#localhost#360#35184372411024       3.256699          162         0   
app1#localhost#390#35184372411024       0.000545            6         0   
app1#localhost#40#35184372411024        2.981486          161         0   
app1#localhost#400#35184372411024       3.331736          159         0   
app1#localhost#404#35184372411024       0.000798            6         0   
app1#localhost#440#35184372411024       2.906523          157         0   
app1#localhost#454#35184372411024       0.000642            6         0   
app1#localhost#46#35184372411024        0.000996            6         0   
app1#localhost#480#35184372411024       2.841209          162         0   
app1#localhost#520#35184372411024       3.185860          159         0   
app1#localhost#536#35184372411024       0.000691            6         0   
app1#localhost#560#35184372411024       3.252610          164         0   
app1#localhost#588#35184372411024       0.000646            6         0   
app1#localhost#600#35184372411024       3.442541          162         0   
app1#localhost#634#35184372411024       0.001628            6         0   
app1#localhost#640#35184372411024       2.886429          158         0   
app1#localhost#654#35184372411024       0.000608            6         0   
app1#localhost#680#35184372411024       2.699609          158         0   
app1#localhost#695#35184372411024       0.000603            6         0   
app1#localhost#720#35184372411024       3.285688          160         0   
app1#localhost#728#35184372411024       0.000552            6         0   
app1#localhost#760#35184372411024       2.745123          161         0   
app1#localhost#794#35184372411024       0.001447            6         0   
app1#localhost#80#35184372411024        3.213426          161

time: 298 ms (started: 2023-03-15 13:03:52 -07:00)


In [20]:
proc_name_hl_view['duration_th'].value_counts()

0.00    28
0.50    22
0.75    10
Name: duration_th, dtype: int64

time: 7.37 ms (started: 2023-03-15 13:04:09 -07:00)


In [22]:
proc_name_hl_view.query('duration_th >= 0.5')

duration_sum  index_count  size_min  \
proc_name                                                                 
app1#localhost#0#35184372411024         4.115100        15028         0   
app1#localhost#1000#35184372411024      3.061782          159         0   
app1#localhost#1040#35184372411024      3.071674          159         0   
app1#localhost#1080#35184372411024      2.932791          159         0   
app1#localhost#1120#35184372411024      2.950001          164         0   
app1#localhost#1160#35184372411024      2.781130          159         0   
app1#localhost#120#35184372411024       2.984170          158         0   
app1#localhost#1200#35184372411024      2.968496          160         0   
app1#localhost#1240#35184372411024      3.239430          158         0   
app1#localhost#160#35184372411024       2.787134          158         0   
app1#localhost#200#35184372411024       3.238349          162         0   
app1#localhost#240#35184372411024       2.829355          160         0   
app1#localhost#280#35184372411024       2.957242          158         0   
app1#localhost#320#35184372411024       3.008698          158         0   
app1#localhost#360#35184372411024       3.256699          162         0   
app1#localhost#40#35184372411024        2.981486          161         0   
app1#localhost#400#35184372411024       3.331736          159         0   
app1#localhost#440#35184372411024       2.906523          157         0   
app1#localhost#480#35184372411024       2.841209          162         0   
app1#localhost#520#35184372411024       3.185860          159         0   
app1#localhost#560#35184372411024       3.252610          164         0   
app1#localhost#600#35184372411024       3.442541          162         0   
app1#localhost#640#35184372411024       2.886429          158         0   
app1#localhost#680#35184372411024       2.699609          158         0   
app1#localhost#720#35184372411024       3.285688          160         0   
app1#localhost#760#35184372411024       2.745123          161         0   
app1#localhost#80#35184372411024        3.213426          161         0   
app1#localhost#800#35184372411024       2.567195          160         0   
app1#localhost#840#35184372411024       3.014022          158         0   
app1#localhost#880#35184372411024       2.948086          158         0   
app1#localhost#920#35184372411024       2.561914          158         0   
app1#localhost#960#35184372411024       2.972840          157         0   

                                    size_max    size_sum  read_time  \
proc_name                                                             
app1#localhost#0#35184372411024     16777216  1298859572   0.001326   
app1#localhost#1000#35184372411024  16777216    16777216   0.000011   
app1#localhost#1040#35184372411024  16777216    16777216   0.000009   
app1#localhost#1080#35184372411024  16777216    16777216   0.000009   
app1#localhost#1120#35184372411024  16777216    16777216   0.000009   
app1#localhost#1160#35184372411024  16777216    16777216   0.000018   
app1#localhost#120#35184372411024   16777216    16777216   0.000011   
app1#localhost#1200#35184372411024  16777216    16777216   0.000008   
app1#localhost#1240#35184372411024  16777216    16777216   0.000010   
app1#localhost#160#35184372411024   16777216    16777216   0.000011   
app1#localhost#200#35184372411024   16777216    16777216   0.000931   
app1#localhost#240#35184372411024   16777216    16777216   0.000009   
app1#localhost#280#35184372411024   16777216    16777216   0.000012   
app1#localhost#320#35184372411024   16777216    16777216   0.000009   
app1#localhost#360#35184372411024   16777216    16777216   0.000010   
app1#localhost#40#35184372411024    16777216    16777216   0.000010   
app1#localhost#400#35184372411024   16777216    16777216   0.000011   
app1#localhost#440#35184372411024   16777216    16777216   0.000008   
app1#localhost#480#35184372411024   16777216    16777216   0.0005

time: 44.6 ms (started: 2023-03-15 13:04:46 -07:00)


In [22]:
file_name_hl_view = bottlenecks[('file_name',)]['high_level_view'].compute()
file_name_hl_view


duration_sum  index_count  \
file_name                                                                       
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.001437            5   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.002457            8   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      2.191869          162   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.003101          130   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.002797          130   
...                                                          ...          ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.002482          130   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.003797          131   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.000537            8   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.002123         6405   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      0.020914            8   

                                                    size_min  size_max  \
file_name                                                                
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  16777216  16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      2055      2055   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0  16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   8454144   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   8454144   
...                                                      ...       ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0   8454144   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0  16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0      6112   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  16777216  16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0    122854   

                                                       size_sum  read_time  \
file_name                                                                    
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...     16777216   0.001307   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         2055   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...     44564480   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      8454144   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      8454144   0.000000   
...                                                         ...        ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      8454144   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...     16908288   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         6112   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  21491613696   0.001981   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...       122854   0.000000   

                                                    write_time  metadata_time  \
file_name                                                                       
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.000000       0.000130   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.000072       0.002385   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.006853       2.191869   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.002899       0.000854   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.002616       0.000726   
...                                                        ...            ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.002297       0.000526   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.003611       0.000729   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.000153       0.000384   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.000000       0.000410   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file.

time: 303 ms (started: 2023-03-16 09:20:21 -07:00)


In [23]:
file_name_hl_view['duration_th'].value_counts()

0.000    710
0.001     27
0.010     15
0.500      2
Name: duration_th, dtype: int64

time: 4.15 ms (started: 2023-03-16 09:20:22 -07:00)


In [24]:
file_name_hl_view.query('duration_th >= 0.5')

duration_sum  index_count  \
file_name                                                                       
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      2.191869          162   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...      2.220052         2528   

                                                    size_min  size_max  \
file_name                                                                
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0  16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...         0      5530   

                                                    size_sum  read_time  \
file_name                                                                 
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  44564480        0.0   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...   1996330        0.0   

                                                    write_time  metadata_time  \
file_name                                                                       
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.006853       2.191869   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...    0.124729       2.095323   

                                                    read_size  write_size  \
file_name                                                                   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...          0    44564480   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...          0     1996330   

                                                    ...  sequential_count  \
file_name                                           ...                     
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  ...                 3   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...  ...               361   

                                                    random_count  close_time  \
file_name                                                                      
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...             0    0.000481   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...             0    0.014551   

                                                    open_time  seek_time  \
file_name                                                                  
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...   2.191843   0.000000   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...   1.984935   0.001064   

                                                    stat_time  duration_pero  \
file_name                                                                      
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...   0.000010       0.532640   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...   0.006172       0.539489   

                                                    duration_perr  \
file_name                                                           
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...       0.281986   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...       0.285612   

                                                    duration_score  \
file_name                                                            
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            high   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...            high   

                                                    duration_th  
file_name                                                        
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...          0.5  
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file...          0.5  

[2 rows x 31 columns]

time: 29.5 ms (started: 2023-03-16 09:20:23 -07:00)


In [25]:
drop_cols = []
groupby = []
for col in file_name_hl_view.columns:
    if 'duration' in col or 'time' in col:
        # drop_cols.append(col)
        pass
    else:
        groupby.append(col)

repeated_behavior = file_name_hl_view \
    .query('duration_th > 0') \
    .drop(columns=drop_cols) \
    .reset_index() \
    .groupby(groupby) \
    .agg({
        'file_name': 'count',
        'duration_sum': 'mean',
        'open_time': 'mean',
    }) \
    .sort_values('file_name', ascending=False) \
    .query('file_name > 1')

repeated_behavior


file_name  \
index_count size_min size_max size_sum read_size write_size metadata_size read_count write_count metadata_count data_count data_size sequential_size random_size sequential_count random_count              
6           49340    49340    49340    0         49340      0             0          1           5              1          49340     49340           0           1                0                    10   
            47144    47144    47144    0         47144      0             0          1           5              1          47144     47144           0           1                0                     6   
163         0        16777216 44564480 0         44564480   0             0          3           160            3          44564480  44564480        0           3                0                     5   
135         0        16777216 84541440 0         84541440   0             0          6           129            6          84541440  84541440        0           6                0                     4   
130         0        8454144  8454144  0         8454144    0             0          1           129            1          8454144   8454144         0           1                0                     2   

                                                                                                                                                                                                duration_sum  \
index_count size_min size_max size_sum read_size write_size metadata_size read_count write_count metadata_count data_count data_size sequential_size random_size sequential_count random_count                 
6           49340    49340    49340    0         49340      0             0          1           5              1          49340     49340           0           1                0                 0.045141   
            47144    47144    47144    0         47144      0             0          1           5              1          47144     47144           0           1                0                 0.014540   
163         0        16777216 44564480 0         44564480   0             0          3           160            3          44564480  44564480        0           3                0                 0.166418   
135         0        16777216 84541440 0         84541440   0             0          6           129            6          84541440  84541440        0           6                0                 0.011092   
130         0        8454144  8454144  0         8454144    0             0          1           129            1          8454144   8454144         0           1                0                 0.022368   

                                                                                                                                                                                                open_time  
index_count size_min size_max size_sum read_size write_size metadata_size read_count write_count metadata_count data_count data_size sequential_size random_size sequential_count random_count             
6           49340    49340    49340    0         49340      0             0          1           5              1          49340     49340           0           1                0              0.044978  
            47144    47144    47144    0         47144      0             0          1           5              1          47144     47144           0           1                0              0.014423  
163         0        16777216 44564480 0         44564480   0             0          3           160            3          44564480  44564480        0           3                0              0.166391  
135         0        16777216 84541440 0         84541440   0             0          6           129            6          84541440  84541440        0           6                0              0.003854  
130         0        8454144  8454144  0         8454144    0             0

time: 36.4 ms (started: 2023-03-16 09:20:23 -07:00)


In [27]:
file_name_ml_view = bottlenecks[('file_name',)]['mid_level_view'].compute()
file_name_ml_view


duration_sum  \
file_name                                          proc_name                                          
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024         0.001437   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024         0.002457   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024         0.007453   
                                                   app1#localhost#1000#35184372411024      2.151749   
                                                   app1#localhost#1040#35184372411024      1.952589   
...                                                                                             ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#996#35184372411024       0.000038   
                                                   app1#localhost#997#35184372411024       0.000036   
                                                   app1#localhost#998#35184372411024       0.000037   
                                                   app1#localhost#999#35184372411024       0.000035   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024         0.020914   

                                                                                       index_count  \
file_name                                          proc_name                                         
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024               5   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024               8   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024               7   
                                                   app1#localhost#1000#35184372411024            5   
                                                   app1#localhost#1040#35184372411024            5   
...                                                                                            ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#996#35184372411024             5   
                                                   app1#localhost#997#35184372411024             5   
                                                   app1#localhost#998#35184372411024             5   
                                                   app1#localhost#999#35184372411024             5   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024               8   

                                                                                       size_min  \
file_name                                          proc_name                                      
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024     16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024         2055   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024     11010048   
                                                   app1#localhost#1000#35184372411024         0   
                                                   app1#localhost#1040#35184372411024         0   
...                                                                                         ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#996#35184372411024   16777216   
                                                   app1#localhost#997#35184372411024   16777216   
                                                   app1#localhost#998#35184372411024   16777216   
                                                   app1#localhost#999#35184372411024   16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024            0   

                                                                                       size_max  \
file_name                        

time: 278 ms (started: 2023-03-16 09:21:03 -07:00)


In [16]:
file_name_ll_view = bottlenecks[('file_name',)]['low_level_view'].compute()
file_name_ll_view


duration_sum  \
file_name                                          proc_name                          trange                 
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    4           0.001437   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    4           0.002457   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    6           0.005920   
                                                                                      7           0.001533   
                                                   app1#localhost#1000#35184372411024 4           0.000003   
...                                                                                                    ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#997#35184372411024  3           0.000036   
                                                   app1#localhost#998#35184372411024  3           0.000037   
                                                   app1#localhost#999#35184372411024  3           0.000035   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    3           0.020722   
                                                                                      4           0.000192   

                                                                                              index_count  \
file_name                                          proc_name                          trange                
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    4                 5   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    4                 8   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    6                 5   
                                                                                      7                 2   
                                                   app1#localhost#1000#35184372411024 4                 1   
...                                                                                                   ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#997#35184372411024  3                 5   
                                                   app1#localhost#998#35184372411024  3                 5   
                                                   app1#localhost#999#35184372411024  3                 5   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    3                 6   
                                                                                      4                 2   

                                                                                              size_min  \
file_name                                          proc_name                          trange             
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    4       16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    4           2055   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411024    6       16777216   
                                                                                      7       11010048   
                                                   app1#localhost#1000#35184372411024 4              0   
...                                                                                                ...   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#997#35184372411024  3       16777216   
                                                   app1#localhost#998#35184372411024  3       16777216   
                                                   app1#localhost#999#35184372411024  3       16777216   
/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_file... app1#localhost#0#35184372411

time: 347 ms (started: 2023-03-16 12:17:15 -07:00)


In [19]:
file_name_ll_view.loc['/p/gpfs1/iopp/temp/cm1r20.3.2846827/config_files/les_ConvBoundLayer/LANDUSE.TBL'].query('size_max > 1')

duration_sum  index_count  size_min  \
proc_name                       trange                                        
app1#localhost#0#35184372411024 4           0.001437            5  16777216   

                                        size_max  size_sum  read_time  \
proc_name                       trange                                  
app1#localhost#0#35184372411024 4       16777216  16777216   0.001307   

                                        write_time  metadata_time  read_size  \
proc_name                       trange                                         
app1#localhost#0#35184372411024 4              0.0        0.00013   16777216   

                                        write_size  ...  sequential_count  \
proc_name                       trange              ...                     
app1#localhost#0#35184372411024 4                0  ...                 1   

                                        random_count  close_time  open_time  \
proc_name                       trange                                        
app1#localhost#0#35184372411024 4                  0    0.000008   0.000114   

                                        seek_time  stat_time  duration_pero  \
proc_name                       trange                                        
app1#localhost#0#35184372411024 4             0.0   0.000008       0.000349   

                                        duration_perr  duration_score  \
proc_name                       trange                                  
app1#localhost#0#35184372411024 4            0.000015            none   

                                        duration_th  
proc_name                       trange               
app1#localhost#0#35184372411024 4               0.0  

[1 rows x 31 columns]

time: 32.5 ms (started: 2023-03-16 12:17:25 -07:00)


In [18]:
for view_key, bottleneck in bottlenecks.items():
    for th in bottleneck.keys():
        print(view_key, th, len(bottleneck[th]))

('trange',) low_level_view 6726
('trange',) mid_level_view 2197
('trange',) high_level_view 393
('file_name',) low_level_view 5427
('file_name',) mid_level_view 3149
('file_name',) high_level_view 754
('proc_name',) low_level_view 4286
('proc_name',) mid_level_view 977
('proc_name',) high_level_view 60
('trange', 'file_name') low_level_view 5427
('trange', 'file_name') mid_level_view 3149
('trange', 'file_name') high_level_view 754
('trange', 'proc_name') low_level_view 4286
('trange', 'proc_name') mid_level_view 977
('trange', 'proc_name') high_level_view 60
('file_name', 'trange') low_level_view 5427
('file_name', 'trange') mid_level_view 2197
('file_name', 'trange') high_level_view 393
('file_name', 'proc_name') low_level_view 4207
('file_name', 'proc_name') mid_level_view 977
('file_name', 'proc_name') high_level_view 60
('proc_name', 'trange') low_level_view 4286
('proc_name', 'trange') mid_level_view 977
('proc_name', 'trange') high_level_view 393
('proc_name', 'file_name') low_l

In [15]:
recorder_analyzer.save_bottlenecks(log_dir=log_dir, bottlenecks=bottlenecks)


time: 4.22 s (started: 2023-03-12 17:31:20 -07:00)


In [16]:
!chmod g+r {log_dir}/bottlenecks/*
!ls -lh {log_dir}/bottlenecks

total 15M
-rw-r--r-- 1 izzet iopp 975K Mar 12 17:31 file_name.json
-rw-r--r-- 1 izzet iopp 1.1M Mar 12 17:31 file_name_proc_name.json
-rw-r--r-- 1 izzet iopp 985K Mar 12 17:31 file_name_proc_name_trange.json
-rw-r--r-- 1 izzet iopp 985K Mar 12 17:31 file_name_trange.json
-rw-r--r-- 1 izzet iopp 1.1M Mar 12 17:31 file_name_trange_proc_name.json
-rw-r--r-- 1 izzet iopp 1.1M Mar 12 17:31 proc_name.json
-rw-r--r-- 1 izzet iopp 975K Mar 12 17:31 proc_name_file_name.json
-rw-r--r-- 1 izzet iopp 985K Mar 12 17:31 proc_name_file_name_trange.json
-rw-r--r-- 1 izzet iopp 985K Mar 12 17:31 proc_name_trange.json
-rw-r--r-- 1 izzet iopp 975K Mar 12 17:31 proc_name_trange_file_name.json
-rw-r--r-- 1 izzet iopp 985K Mar 12 17:31 trange.json
-rw-r--r-- 1 izzet iopp 975K Mar 12 17:31 trange_file_name.json
-rw-r--r-- 1 izzet iopp 1.1M Mar 12 17:31 trange_file_name_proc_name.json
-rw-r--r-- 1 izzet iopp 1.1M Mar 12 17:31 trange_proc_name.json
-rw-r--r-- 1 izzet iopp 975K Mar 12 17:31 trange_proc_name_fil